score.go

```diff
@@ -0,0 +1,952 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type PeerScoreParams struct {
+	// Score parameters per topic.
+	Topics map[string]*TopicScoreParams
+
+	// P5: Application-specific peer scoring
+	AppSpecificScore  func(p peer.ID) float64
+	AppSpecificWeight float64
+
+	// P6: IP-colocation factor.
+	// The parameter has an associated counter which counts the number of peers with the same IP.
+	// If the number of peers in the same IP exceeds IPColocationFactorThreshold, then the value
+	// is the square of the difference, ie (PeersInSameIP - IPColocationThreshold)^2.
+	// If the number of peers in the same IP is less than the threshold, then the value is 0.
+	// The weight of the parameter MUST be negative, unless you want to disable for testing.
+	// Note: In order to simulate many IPs in a managable manner when testing, you can set the weight to 0
+	//       thus disabling the IP colocation penalty.
+	IPColocationFactorWeight    float64
+	IPColocationFactorThreshold int
```

Comment:
Note: this doesn't work well for IPv6 at all so I'm not sure how useful it is in general. I assume an attacker would just disable IPv4 and pick IPv6 addresses out of their ISP assigned address range at will. ISPs are supposed to give out 64bit ranges.

Reply:
We can use `/64`s for IPv6 addresses, that's an easy fix.

Reply:
In general it is _very_ useful as it makes it very hard to launch a sybil attack using just a few IP addresses (which is something that has been observed in the filecoin testnet).

Reply:
I think we need to track both the IP and the /64 in the case of IPv6.
We can also add a whitelist for eg datacenters, so that those /64 don't get penalized.

Reply:
I wonder if a better approach would be to simply say that scores "bleed" between peers with similar IP addresses. That would allow us to say things like "everyone in this IP block appears to be bad, let's reduce their collective score".

Of course, that could be an attack vector...

Reply:
Implemented the `/64` and whitelisting for IPv6.